app/models/mixins/miq_provision_quota_mixin.rb

```diff
@@ -296,33 +315,80 @@ def quota_provision_stats(prov_method, options)
              }
 
     send(prov_method, options).each do |pr|
-      num_vms_for_request = pr.get_option(:number_of_vms).to_i
-      if options[:nil_vm_id_only] == true && pr.miq_request_tasks.length == num_vms_for_request
-        no_vm = pr.miq_request_tasks.find_all { |p| p.destination_id.nil? && p.state != 'finished' }
-        num_vms_for_request = no_vm.length
+      num_vms_for_request = number_of_vms(pr)
+      next if num_vms_for_request.zero?
+      flavor_obj = flavor(pr)
+      result[:count] += num_vms_for_request
+      result[:memory] += memory(pr, cloud?(pr), vendor(pr), flavor_obj) * num_vms_for_request
+      result[:cpu] += number_of_cpus(pr, cloud?(pr), flavor_obj) * num_vms_for_request
+      result[:storage] += storage(pr, cloud?(pr), vendor(pr), flavor_obj) * num_vms_for_request
+      result[:ids] << pr.id
+
+      pr.miq_request_tasks.each do |p|
+        next unless p.state == 'Active'
+        host_id, storage_id = p.get_option(:dest_host).to_i, p.get_option(:dest_storage).to_i
+        active = result[:active]
+        active[:memory_by_host_id][host_id] += memory(p, cloud?(pr), vendor(pr), flavor_obj)
+        active[:cpu_by_host_id][host_id] += number_of_cpus(p, cloud?(pr), flavor_obj)
+        active[:storage_by_id][storage_id] += storage(p, cloud?(pr), vendor(pr), flavor_obj)
+        active[:vms_by_storage_id][storage_id] << p.id
+        active[:ids] << p.id
       end
+    end
+    result
+  end
 
-      unless num_vms_for_request.zero?
-        new_disk_storage_size = pr.get_new_disks.inject(0) { |s, d| s += d[:sizeInMB].to_i } * 1.megabyte
-        result[:count] += num_vms_for_request
-        result[:memory] += pr.get_option(:vm_memory).to_i * num_vms_for_request
-        result[:cpu] += pr.get_option(:number_of_cpus).to_i * num_vms_for_request
-        result[:storage] += (pr.vm_template.allocated_disk_storage.to_i + new_disk_storage_size) * num_vms_for_request
-        result[:ids] << pr.id
-
-        # Include a resource breakdown for actively provisioning records
-        pr.miq_request_tasks.each do |p|
-          next unless p.state == 'active'
-          host_id, storage_id = p.get_option(:dest_host).to_i, p.get_option(:dest_storage).to_i
-          active = result[:active]
-          active[:memory_by_host_id][host_id] += p.get_option(:vm_memory).to_i
-          active[:cpu_by_host_id][host_id] += p.get_option(:number_of_cpus).to_i
-          active[:storage_by_id][storage_id] += p.vm_template.allocated_disk_storage.to_i + new_disk_storage_size
-          active[:vms_by_storage_id][storage_id] << p.id
-          active[:ids] << p.id
-        end
+  def number_of_vms(request)
+    num_vms_for_request = request.get_option(:number_of_vms).to_i
+    if options[:nil_vm_id_only] == true && request.miq_request_tasks.length == num_vms_for_request
+      no_vm = request.miq_request_tasks.find_all { |p| p.destination_id.nil? && p.state != 'finished' }
```

Comment:
I believe `request.miq_request_tasks` is an ActiveRecord Collection, correct?
If so, prefer `where` over `find_all`.
`request.miq_request_tasks.where(:destination_id => nil).where.not(:state => 'finished')`